File: bedrock/src/multiplayer/logic/solvency.test.ts
import { describe, expect, it } from 'vitest';
import { type BankrollState, bankrollAvailable, settleBankroll, tryReserve } from '../../core/logic/ledger';
import { seededRng } from '../../core/logic/rng';
import { UNKNOWN_WORST_CASE_PER_CHIP, cheapestWorstCase, isBroke, routeHouseResult, worstCaseAt, worstCasePerChip } from './solvency';

describe('worst case table', () => {
  it('matches GAME_DESIGN §18.2 examples', () => {
    expect(worstCasePerChip('roulette')).toBe(36);
    expect(worstCasePerChip('blackjack')).toBe(17.5);
    expect(worstCasePerChip('craps')).toBe(13);
    expect(worstCasePerChip('plinko')).toBe(170);
    expect(worstCasePerChip('slots', 'copper')).toBe(150);
    expect(worstCasePerChip('slots', 'netherite')).toBe(1000);
    expect(worstCasePerChip('slots', 'unknown_variant')).toBe(1000);
    expect(worstCasePerChip('poker')).toBe(0);
    expect(worstCasePerChip('mystery')).toBe(UNKNOWN_WORST_CASE_PER_CHIP);
  });
  it('game-provided overrides win', () => {
    expect(worstCasePerChip('roulette', undefined, { roulette: 40 })).toBe(40);
    expect(worstCasePerChip('slots', 'copper', { 'slots.copper': 90 })).toBe(90);
  });
  it('rounds up to whole chips', () => {
    expect(worstCaseAt(17.5, 1)).toBe(18);
    expect(worstCaseAt(17.5, 2)).toBe(35);
    expect(worstCaseAt(0, 100)).toBe(0);
  });
});

describe('insolvency', () => {
  const tables = [
    { game: 'roulette', minBet: 5, open: true }, // 180
    { game: 'blackjack', minBet: 10, open: true }, // 175
    { game: 'coin_flip', minBet: 1, open: false }, // closed: ignored
    { game: 'poker', minBet: 1, open: true }, // no exposure
  ];
  it('uses the cheapest open exposed table', () => {
    expect(cheapestWorstCase(tables)).toBe(175);
    expect(isBroke({ balance: 174, reserved: 0 }, tables)).toBe(true);
    expect(isBroke({ balance: 175, reserved: 0 }, tables)).toBe(false);
  });
  it('a casino with only poker or no tables is never broke', () => {
    expect(isBroke({ balance: 0, reserved: 0 }, [{ game: 'poker', minBet: 1, open: true }])).toBe(false);
    expect(isBroke({ balance: 0, reserved: 0 }, [])).toBe(false);
  });
  it('minimum bet below 1 counts as 1', () => {
    expect(cheapestWorstCase([{ game: 'coin_flip', minBet: 0, open: true }])).toBe(2);
  });
});

describe('fallback routing', () => {
  it('house wins go to the bankroll', () => {
    expect(routeHouseResult({ balance: 0, reserved: 0 }, 100, 0)).toEqual({ bankrollDelta: 100, bankCovered: 0 });
    expect(routeHouseResult({ balance: 0, reserved: 0 }, 100, 100)).toEqual({ bankrollDelta: 0, bankCovered: 0 });
  });
  it('house losses come out of the unreserved bankroll only', () => {
    expect(routeHouseResult({ balance: 500, reserved: 0 }, 100, 200)).toEqual({ bankrollDelta: -100, bankCovered: 0 });
    expect(routeHouseResult({ balance: 500, reserved: 450 }, 100, 200)).toEqual({ bankrollDelta: -50, bankCovered: 50 });
  });
});

/**
 * Monte-Carlo: an owner banking honest games earns the house edge on the handle, and the
 * reservation rule keeps the bankroll solvent (never negative, reservations ≤ balance).
 */
describe('Monte-Carlo: owner earns the house edge', () => {
  function simulate(game: 'roulette' | 'coin_flip', rounds: number, seed: number) {
    const rng = seededRng(seed);
    let b: BankrollState = { balance: 100_000, reserved: 0 };
    const start = b.balance;
    const bet = 25; // 25 × 1.96 = 49: whole-chip coin-flip payouts, so the edge is exact
    let handle = 0;
    let refused = 0;
    let violations = 0;
    for (let i = 0; i < rounds; i++) {
      const wc = worstCaseAt(worstCasePerChip(game), bet);
      const r = tryReserve(b, wc);
      if (!r) {
        refused++;
        continue;
      }
      b = r;
      let ret: number;
      if (game === 'roulette') ret = Math.floor(rng.next() * 37) === 17 ? bet * 36 : 0; // straight-up 35:1
      else ret = rng.next() < 0.5 ? Math.floor(bet * 1.96) : 0; // coin flip pays 0.96:1
      if (ret > wc) violations++;
      b = settleBankroll(b, { reserved: wc, stake: bet, payout: ret });
      handle += bet;
      if (b.balance < 0 || b.reserved !== 0) violations++;
    }
    return { edge: (b.balance - start) / handle, refused, violations };
  }

  it('roulette straight-up: owner edge ≈ 1/37 = 2.70 %', () => {
    const { edge, refused, violations } = simulate('roulette', 2_000_000, 7);
    expect(refused).toBe(0);
    expect(violations).toBe(0);
    // σ ≈ 5.84 / √2e6 ≈ 0.41 %; tolerance ≈ 3.6σ
    expect(Math.abs(edge - 1 / 37)).toBeLessThan(0.015);
  });

  it('coin flip: owner edge = 2 %', () => {
    const { edge, violations } = simulate('coin_flip', 1_000_000, 11);
    expect(violations).toBe(0);
    // σ ≈ 0.98 / √1e6 ≈ 0.1 %
    expect(Math.abs(edge - 0.02)).toBeLessThan(0.004);
  });

  it('a tiny bankroll refuses bets it cannot cover and never goes negative', () => {
    const rng = seededRng(3);
    let b: BankrollState = { balance: 100, reserved: 0 };
    let accepted = 0;
    let negative = 0;
    for (let i = 0; i < 10_000; i++) {
      const bet = 1 + Math.floor(rng.next() * 10);
      const wc = worstCaseAt(36, bet);
      const r = tryReserve(b, wc);
      if (!r) {
        expect(bankrollAvailable(b)).toBeLessThan(wc);
        continue;
      }
      accepted++;
      b = settleBankroll(r, { reserved: wc, stake: bet, payout: Math.floor(rng.next() * 37) === 0 ? bet * 36 : 0 });
      if (b.balance < 0) negative++;
    }
    expect(negative).toBe(0);
    expect(accepted).toBeGreaterThan(0);
  });
});

File: java/src/test/java/dev/nezo/burmaldaholic/multiplayer/logic/SolvencyTest.java
package dev.nezo.burmaldaholic.multiplayer.logic;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import dev.nezo.burmaldaholic.core.economy.AccountId;
import dev.nezo.burmaldaholic.core.economy.Economy;
import dev.nezo.burmaldaholic.core.economy.Ledger;
import java.util.List;
import java.util.SplittableRandom;
import java.util.UUID;
import org.junit.jupiter.api.Test;

class SolvencyTest {
	@Test
	void worstCaseTableMatchesTheSpecExamples() {
		assertEquals(36, Solvency.worstCasePerChip("roulette", "roulette_table"));
		assertEquals(17.5, Solvency.worstCasePerChip("blackjack", "blackjack_table"));
		assertEquals(13, Solvency.worstCasePerChip("craps", null));
		assertEquals(500, Solvency.worstCasePerChip("slots", "slot_machine_copper"));
		assertEquals(2000, Solvency.worstCasePerChip("slots", "slot_machine_gold"));
		assertEquals(5000, Solvency.worstCasePerChip("slots", "slot_machine_netherite"));
		assertEquals(5000, Solvency.worstCasePerChip("slots", "unknown_machine"));
		assertEquals(170, Solvency.worstCasePerChip("extras", "plinko_machine"));
		assertEquals(0, Solvency.worstCasePerChip("poker", "poker_table"));
		assertEquals(Solvency.UNKNOWN_WORST_CASE_PER_CHIP, Solvency.worstCasePerChip("mystery", "mystery_table"));
	}

	@Test
	void worstCaseRoundsUpToWholeChips() {
		assertEquals(18, Solvency.worstCaseAt(17.5, 1));
		assertEquals(35, Solvency.worstCaseAt(17.5, 2));
		assertEquals(0, Solvency.worstCaseAt(0, 100));
		assertEquals(0, Solvency.worstCaseAt(36, 0));
	}

	@Test
	void insolvencyUsesTheCheapestOpenExposedTable() {
		List<Solvency.ExposedTable> tables = List.of(
			new Solvency.ExposedTable("roulette", "roulette_table", 5, true),      // 180
			new Solvency.ExposedTable("blackjack", "blackjack_table", 10, true),   // 175
			new Solvency.ExposedTable("extras", "wheel_of_fortune", 1, false),     // closed: ignored
			new Solvency.ExposedTable("poker", "poker_table", 1, true));           // no exposure
		assertEquals(175, Solvency.cheapestWorstCase(tables).orElseThrow());
		assertTrue(Solvency.isBroke(174, tables));
		assertFalse(Solvency.isBroke(175, tables));
	}

	@Test
	void casinoWithoutExposureIsNeverBroke() {
		assertFalse(Solvency.isBroke(0, List.of(new Solvency.ExposedTable("poker", "poker_table", 1, true))));
		assertFalse(Solvency.isBroke(0, List.of()));
		assertEquals(18, Solvency.cheapestWorstCase(List.of(new Solvency.ExposedTable("blackjack", null, 0, true))).orElseThrow(),
			"minimum bet below 1 counts as 1");
	}

	@Test
	void withdrawalsAreLimitedToTheUnreservedBankroll() {
		assertEquals(400, Solvency.withdrawable(1000, 600));
		assertEquals(0, Solvency.withdrawable(500, 600));
		assertEquals(500, Solvency.withdrawable(500, 0));
	}

	// ---- Monte-Carlo with core's real ledger (same legs as CasinoTableBlockEntity#placeBet / #settle) -------

	private static final long MAX = 1_000_000_000_000L;
	private static final String BANK = "multiplayer:charter/c1";

	private record Sim(double edge, int refused, int violations) {}

	/** Plays rounds against an owner bankroll: reserve worst case, stake → bankroll, payout ← bankroll, release. */
	private static Sim simulate(String game, int rounds, long bankroll, long bet, long seed) {
		Ledger ledger = new Ledger();
		UUID owner = UUID.randomUUID();
		UUID player = UUID.randomUUID();
		ledger.openBankroll(BANK, owner);
		ledger.commit(List.of(new Ledger.Leg(AccountId.bankroll(BANK), bankroll)), MAX, null);
		SplittableRandom rng = new SplittableRandom(seed);
		long handle = 0;
		int refused = 0;
		int violations = 0;
		for (int i = 0; i < rounds; i++) {
			ledger.setBalance(player, 1_000_000, MAX);
			long wc = Solvency.worstCaseAt(Solvency.worstCasePerChip(game, null), bet);
			if (!ledger.reserve(BANK, wc)) {
				refused++;
				continue;
			}
			assertTrue(ledger.commit(List.of(new Ledger.Leg(AccountId.player(player), -bet), new Ledger.Leg(AccountId.bankroll(BANK), bet)), MAX, null).ok());
			long payout = switch (game) {
				case "roulette" -> rng.nextInt(37) == 17 ? bet * 36 : 0;          // straight-up 35:1
				case "coin_flip" -> rng.nextBoolean() ? (long) Math.floor(bet * 1.96) : 0;
				default -> throw new IllegalArgumentException(game);
			};
			if (payout > wc) {
				violations++;
			}
			ledger.release(BANK, wc);
			if (payout > 0 && !ledger.commit(List.of(new Ledger.Leg(AccountId.bankroll(BANK), -payout), new Ledger.Leg(AccountId.player(player), payout)), MAX, null).ok()) {
				violations++;
			}
			Economy.BankrollInfo info = ledger.bankroll(BANK).orElseThrow();
			if (info.balance() < 0 || info.reserved() != 0) {
				violations++;
			}
			handle += bet;
		}
		long end = ledger.bankroll(BANK).orElseThrow().balance();
		return new Sim((double) (end - bankroll) / handle, refused, violations);
	}

	@Test
	void ownerEarnsTheRouletteEdge() {
		Sim s = simulate("roulette", 2_000_000, 100_000, 25, 7);
		assertEquals(0, s.refused());
		assertEquals(0, s.violations());
		// edge 1/37 = 2.70 %; σ ≈ 5.84 / √2e6 ≈ 0.41 % → tolerance ≈ 3.6σ
		assertEquals(1.0 / 37, s.edge(), 0.015);
	}

	@Test
	void ownerEarnsTheCoinFlipEdge() {
		Sim s = simulate("coin_flip", 1_000_000, 100_000, 25, 11);
		assertEquals(0, s.violations());
		// 25 × 1.96 = 49 whole chips: exact 2 % edge; σ ≈ 0.1 %
		assertEquals(0.02, s.edge(), 0.004);
	}

	@Test
	void tinyBankrollRefusesUncoverableBetsAndNeverGoesNegative() {
		Sim s = simulate("roulette", 20_000, 100, 5, 3);
		assertEquals(0, s.violations());
		assertTrue(s.refused() > 0, "a 100-chip bankroll cannot cover 180-chip exposures");
	}
}
